rfcs/react-components/convergence/custom-styling.md

```diff
@@ -0,0 +1,476 @@
+# RFC: Custom Styling
+
+## Contributors
+
+- Geoff Cox
+- Mason Tejera - MADS use cases
+- Ben Howell - easier recomposition
+- Micah Godbolt - component CSS vars
+
+## Summary
+
+Fluent UI React v9 is built today to adopt the Fluent 2 design as its default style. We strongly recommend that
+Microsoft applications leverage the default style and limit customization to a brand-based theme.
+
+Currently, v9 has mechanisms to support several customization scenarios:
+
+- Custom themes => Customize style values for all components
+- `classname` prop => Customize the style of one instance of a component
+- Extended design tokens => Support theme for a custom component or custom class/style
+- Hooks recomposition => Create a new component customizing the behavior, style, or rendering of an existing component.
+
+See the appendix for more detailed analysis of the capabilities and limitations of each of these mechansisms.
+
+## Problem
+
+Partners migrating from v0/v8 to v9 have brought up that v9's custom styling does not provide the same granularity of
+customization as previous versions. See the appendix for detailed partner use cases and scenarios.
+
+Partners producing component libraries or component design surfaces are the most concerned about the lack of
+customization. They would like to customize the style of a single type of component without changing the style of
+other components.
+
+Example: Change the border radius and border width of all Button components without changing those styles for
+Input components.
+
+## Option A: React.Context useCustomStyles hooks
+
+Each component could check context for a useCustomStyle hook within the useComponentStyle hook.
+If this hook was defined, it is called at the end of the method before returning state.
+
+> These example use Component as a substitute for a component like Button or Avatar.
+
+```tsx
+const customStyles = useContextSelector(ComponentCustomStylesContext, context => context?.useCustomStyles);
+
+//...component styling here
+
+return customStyles ? customStyles(state) : state;
+```
+
+Each component defines own types for props and state, so would need to define its own context value types,
+context, and context provider.
+
+```tsx
+export type ComponentCustomStylesContextValue = {
+  useCustomStyles?: (state: ComponentState) => ComponentState;
+};
+
+export const ComponentCustomStylesContext: Context<ComponentCustomStylesContextValue> = createContext<ComponentCustomStylesContextValue>(
+  {},
+);
+
+export const ComponentCustomStylesContextProvider = ComponentCustomStylesContext.Provider;
```

Comment:
Do I correctly get that it's a provider per each component? I.e. `ButtonCustomStylesContextProvider`, `AvatarCustomStylesContextProvider` and etc.?

```jsx
function App() {
  return (
    <ButtonCustomStylesContextProvider value={{ useCustomStyles }}>
      <Button />
    </ButtonCustomStylesContextProvider>
  );
}
```

- There is no point to use `useContextSelector()` as `ComponentCustomStylesContextValue` could be just a function: it will also be a small performance improvement as `useContextSelector()` is not really free
- It will not work great if there are multiple versions of a `@fluentui/react-*` package ([problem with context singletons](https://github.com/microsoft/fluentui/tree/master/packages/react-components/global-context)) i.e. customizations may not be applied to some components

Reply:
Writing so many providers is one of the things that makes this option unwieldy for consumers and likely will be discarded in favor of something more lightweight. Understood (I think) about the context selector.